### PR TITLE
Gui: allow LineEdit to grow with text - fixes #17747

### LINKED ISSUE #17747: UI: Tree view renaming field does not automatically expand and the default can be very small
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Hello everyone,

Here is a maybe easy to fix little thing. The rename entry field object is apparently based on the length of the text, not the available size in the window. And it does not resize when entering more info.

Meaning if the old name is short, and the new name is long, it gets cut off very fast and looks a bit silly. 
 
image
image

This probably doesn't affect anything. Definitely a nice to have polish issue.

### Full version info

```shell
OS: Ubuntu 24.04.1 LTS (XFCE/xubuntu/xcb)
Architecture: x86_64
Version: 1.1.0dev.39104 +2 (Git)
Build type: Unknown
Branch: AboutCopyConfirm
Hash: fb16a9d376abdadbbedbdc8dc1afdf076198e443
Python 3.12.3, Qt 5.15.13, Coin 4.0.2, Vtk 9.1.0, OCC 7.6.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Fusion
Installed mods: 
  * lattice2 1.0.0
```

### Subproject(s) affected?

None

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by VincidaB**
This adds a `DynamicQLineEdit` dynamic class for the model Tree so that the edit field can grow with what is being typed during renaming. This fixes #17747 .

https://github.com/user-attachments/assets/615bcedf-70ba-4437-a2b3-51b77f897d21

I am not sure if this merits an new class or if it should have been added to `ExpLineEdit`.  (I didn't want to disturb the existing class).

### CODE REVIEW

File: src/Gui/Tree.cpp [hunk 1]
```diff
@@ -540,6 +540,31 @@ void TreeWidgetItemDelegate::initStyleOption(QStyleOptionViewItem *option,
...
+{
+public:
+    DynamicQLineEdit(QWidget *parent = nullptr) : ExpLineEdit(parent) {}
+
+    QSize sizeHint() const override
+    {
+        QSize size = QLineEdit::sizeHint(); 
+        QFontMetrics fm(font());
+        int availableWidth = parentWidget()->width() - geometry().x() - 20; // Calculate available width
+        size.setWidth(std::min(fm.horizontalAdvance(text()) + 30, availableWidth)); // Add some padding and max width
```
Comment (FlachyJoe) [inline src/Gui/Tree.cpp:553]: Where are `20` and `30` come from?

File: src/Gui/Tree.cpp [hunk 2]
```diff
@@ -540,6 +540,31 @@ void TreeWidgetItemDelegate::initStyleOption(QStyleOptionViewItem *option,
...
+        Base::Console().Warning("sizeHint %d %d\n", size.width(), size.height());
```
Comment (FlachyJoe) [inline src/Gui/Tree.cpp:554]: Shouldn't this be a `Log` instead of a `Warning`?

Comment (VincidaB) [reply src/Gui/Tree.cpp:554]: There probably shouldn't be any log at all actually, I'll remove it.

File: src/Gui/Tree.cpp (hunk 1 above)
Comment (VincidaB) [reply src/Gui/Tree.cpp:553]: Some padding as to not overflow on the right, but might not be necessary.
I'll remove it and test.

Comment (VincidaB) [reply src/Gui/Tree.cpp]: I changed the 30 to : 
```c++
2 * (style()->pixelMetric(QStyle::PM_FocusFrameHMargin) + 1)
                    + 2 * style()->pixelMetric(QStyle::PM_LayoutHorizontalSpacing)
                    + TreeParams::getItemBackgroundPadding();
```
I took inspiration `QRect TreeWidgetItemDelegate::calculateItemRect()`, but I am not certain of what I did.

**Comment by maxwxyz** [comment]
@VincidaB ping for feedback

**Comment by VincidaB** [comment]
I changed the hard coded value for the width offset of the text field. This should be good

File: src/Gui/Tree.cpp (hunk 1 above)
Comment (kadet1090) [reply src/Gui/Tree.cpp]: Not really as this is something that was mostly a result of trial and error. How it really should look like:

```c++
style()->pixelMetric(QStyle::PM_PaddingLeft) + style()->pixelMetric(QStyle::PM_PaddingRight) + 
style()->pixelMetric(QStyle::PM_BorderLeft) + style()->pixelMetric(QStyle::PM_BorderRight) + 
TreeParams::getItemBackgroundPadding();
```

However, Qt does not expose ability to read padding or border. From my experiments I concluded that to achieve somewhat reliable spacing default horizontal margins of QFrames combined with default layout horizontal padding can be used to have good enough results.

**Review by kadet1090 — APPROVED** [review]

**Pull request merged**